File: frontend/src/api/schedule.ts
import apiClient from './client'
import type {
  CalendarVisibility,
  Page,
} from '@/types'

export interface ScheduleDto {
  id: string
  content: string
  description: string
  position: number
  year: number
  month: number
  dayOfMonth: number
  startDateTime: string
  endDateTime: string
  isTagged: boolean
  owner: string
  taggedByMember?: ScheduleTagMemberDto | null
  tags: ScheduleTagMemberDto[]
  visibility?: CalendarVisibility
  dateToCompare: string
  attachments: ScheduleAttachmentDto[]
  // Computed fields from backend
  startDate: string
  daysFromStart: number
  endDate: string
  totalDays: number
  curDate: string
}

export interface ScheduleTagMemberDto {
  id: number | null
  name: string
  email?: string | null
  teamId?: number | null
  team?: string | null
  hasProfilePhoto?: boolean
  profilePhotoVersion?: number
}

export interface ScheduleAttachmentDto {
  id: string
  originalFilename: string
  contentType: string
  size: number
  hasThumbnail: boolean
  thumbnailUrl: string | null
  orderIndex: number
}

export interface ScheduleSaveDto {
  id?: string
  memberId: number
  content: string
  description?: string
  visibility?: CalendarVisibility
  startDateTime: string
  endDateTime: string
  tagFriendIds?: number[]
  attachmentSessionId?: string
  orderedAttachmentIds?: string[]
  aiTimeParsingRequested?: boolean
}

export interface ScheduleSearchResult {
  id: string
  content: string
  description: string
  startDateTime: string
  endDateTime: string
  hasAttachments: boolean
}

export interface ScheduleBasicInfo {
  id: string
  memberId: number
  memberName: string
  startDateTime: string
  content: string
}

export const scheduleApi = {
  getScheduleById: async (scheduleId: string): Promise<ScheduleBasicInfo> => {
    const response = await apiClient.get<ScheduleBasicInfo>(`/schedules/${scheduleId}`)
    return response.data
  },

  // The response is indexed by zero-based day of month for the calendar view.
  getSchedules: async (
    memberId: number,
    year: number,
    month: number
  ): Promise<ScheduleDto[][]> => {
    const response = await apiClient.get<ScheduleDto[][]>('/schedules', {
      params: { memberId, year, month },
    })
    return response.data
  },

  searchSchedules: async (
    memberId: number,
    query: string,
    page: number = 0,
    size: number = 10
  ): Promise<Page<ScheduleSearchResult>> => {
    const response = await apiClient.get<Page<ScheduleSearchResult>>(
      `/schedules/${memberId}/search`,
      {
        params: { q: query, page, size },
      }
    )
    return response.data
  },

  saveSchedule: async (schedule: ScheduleSaveDto): Promise<{ id: string }> => {
    const response = await apiClient.post<{ id: string }>('/schedules', schedule)
    return response.data
  },

  deleteSchedule: async (scheduleId: string): Promise<void> => {
    await apiClient.delete(`/schedules/${scheduleId}`)
  },

  reorderSchedulePositions: async (scheduleIds: string[]): Promise<void> => {
    await apiClient.patch('/schedules/positions', scheduleIds)
  },

  untagSelf: async (scheduleId: string): Promise<void> => {
    await apiClient.delete(`/schedules/${scheduleId}/tags`)
  },
}

File: ios/Dutypark/Core/Models/Schedule.swift
import Foundation

enum CalendarVisibility: String, Decodable, Sendable {
    case `public` = "PUBLIC"
    case friends = "FRIENDS"
    case family = "FAMILY"
    case `private` = "PRIVATE"

    var displayName: String {
        switch self {
        case .public: return "전체 공개"
        case .friends: return "친구에게만"
        case .family: return "가족에게만"
        case .private: return "비공개"
        }
    }

    var iconName: String {
        switch self {
        case .public: return "globe"
        case .friends: return "person.2"
        case .family: return "house"
        case .private: return "lock"
        }
    }
}

struct Schedule: Decodable, Identifiable, Sendable {
    let id: String
    let content: String
    let description: String?
    let position: Int
    let year: Int
    let month: Int
    let dayOfMonth: Int
    let startDateTime: String
    let endDateTime: String
    let isTagged: Bool
    let owner: String?
    let tags: [ScheduleTag]
    let visibility: CalendarVisibility?
    let dateToCompare: String?
    let attachments: [Attachment]
    let daysFromStart: Int?
    let totalDays: Int?

    // Computed property to extract time from startDateTime
    var startTime: String? {
        guard startDateTime.contains("T") else { return nil }
        let parts = startDateTime.split(separator: "T")
        guard parts.count == 2 else { return nil }
        let timePart = String(parts[1].prefix(5)) // "HH:mm"
        return timePart == "00:00" ? nil : timePart
    }

    // Computed property to extract time from endDateTime
    var endTime: String? {
        guard endDateTime.contains("T") else { return nil }
        let parts = endDateTime.split(separator: "T")
        guard parts.count == 2 else { return nil }
        let timePart = String(parts[1].prefix(5)) // "HH:mm"
        return timePart == "00:00" ? nil : timePart
    }
}

struct ScheduleTag: Decodable, Identifiable, Sendable {
    var id: Int { memberId }
    let memberId: Int
    let memberName: String
}

struct ScheduleListResponse: Decodable, Sendable {
    let schedules: [Schedule]
}

// MARK: - Requests
struct CreateScheduleRequest: Encodable {
    let content: String
    let description: String?
    let year: Int
    let month: Int
    let dayOfMonth: Int
    let startDateTime: String
    let endDateTime: String
    let visibility: String?
    let taggedMemberIds: [Int]?
    let attachmentSessionId: String?
    let orderedAttachmentIds: [String]?
}

struct UpdateScheduleRequest: Encodable {
    let content: String
    let description: String?
    let year: Int
    let month: Int
    let dayOfMonth: Int
    let startDateTime: String
    let endDateTime: String
    let visibility: String?
    let taggedMemberIds: [Int]?
    let attachmentSessionId: String?
    let orderedAttachmentIds: [String]?
}
